### ISSUE TITLE
Strip trailing slash doesn't take into account query or hash in the url

### ISSUE THREAD

**Issue by diestrin** [angular/angular#14905]
**I'm submitting a ...**
```
[x] bug report
```

**Current behavior**
The Location class is removing the trailing slash, but it's just looking at the end of the string and the string might contain the query and hash values. Currently if you have a a trailing slash and a query parameter, the trailing slash is not being removed.

**Expected behavior**
The Location::stripTrailingSlash should take care of remove the trailing slash even if the query or hash are present.

**Minimal reproduction of the problem with instructions**
Given the route definition `{ path: '/test', pathMatch: 'full' }`
This is the result:
```
/test => Works
/test/ => Works, rewrites browser url to /test
/test/?a=b => Error, doesn't find the route
```

**What is the motivation / use case for changing the behavior?**
It's a bug and causes routes not to work properly

* **Angular version:** 2.1.2
Checked the regex in master branch and it's still the same as in 2.1.2

* **Browser:** [all]
 
* **Language:** [all]

**Comment by diestrin** [comment]
I was able to fix this temporally in my code by overriding the function. Leaving the code here in case anyone else has the same issue:
```ts
import { Location } from '@angular/common';

const trailingSlash = /\/$|\/(?=\?)|\/(?=#)/g;

Location.stripTrailingSlash = (url: string) => url.replace(trailingSlash, '');
```

**Comment by DzmitryShylovich** [comment]
pls add a plunkr

**Comment by pxwise** [comment]
This is a problem for us migrating legacy pages which require trailing slashes in universal and may also receive query params from inbound ad campaign links.

This problem is compounded when two modules share the same base path but one also accepts :input args.

For instance ModuleA uses route `blog` for a landing page and ModuleB uses `blog/:title` for an individual post.

`blog/` correctly displays the ModuleA landing page.
`blog/?awesome_ad_tracker=taco` will trigger ModuleB and display a broken template.

I can confirm this patch resolves our issue.

**Comment by wushawn** [comment]
Anything else required to land this location trailing slash fix?

**Comment by jasonaden** [comment]
Fixed with #17147
